[stable/traefik] Ability to add arbitrary Secrets and ConfigMaps

jacobsvante opened the pull request:
**What this PR does / why we need it**:

Google Cloud DNS is configured differently from all other ACME DNS providers in that it expects a file to be available inside the container (because of how the underlying `lego` library is set up). This didn’t map well to how helm packages are expected to be provided with configuration/secrets.

To support this the ability to add any number of Secrets and ConfigMaps has been added.

~Implementation details:~
~The contents of `GCE_SERVICE_ACCOUNT_FILE_DATA` will be made available in a file inside the container and its path will be set as the value of environment variable `GCE_SERVICE_ACCOUNT_FILE`, which the [underlying ACME library](https://github.com/xenolf/lego) looks for.~

~That means that any value set manually to `GCE_SERVICE_ACCOUNT_FILE` will be ignored with this change. My general understanding of k8s/helm is quite fresh so I'm not sure if anyone's been able to use the setting at all before without resorting to something like what @Azuka suggested in https://github.com/kubernetes/charts/pull/1299#issuecomment-361450527.~

**Which issue this PR fixes**: fixes #5617

@krancour @emilevauge @dtomcej @ldez

jacobsvante commented:
/assign @krancour

timstoop commented on stable/traefik/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: traefik
-version: 1.29.2
+version: 1.29.4
```
You're adding in a new feature, please bump the minor version of the chart.

jacobsvante replied:
@timstoop Done

timstoop commented:
Looks like another PR for traefik got accepted first. Please merge from master and check the Chart version again.

jacobsvante commented:
@timstoop Merged

timstoop commented:
You're still adding a new feature, so I would expect you to bump the minor version :) To 1.31.0 this time.

timstoop commented:
/ok-to-test

jacobsvante commented:
Please hold with merging this PR, it looks as if the env vars aren't properly set.

timstoop commented:
/hold

jacobsvante commented:
Okay, fixed the problem. Should be good to merge.

/hold cancel

krancour commented:
I can easily imagine many other configuration options (if not in the present, then in the future) where a file might need to be present on the system, so I'd like to suggest a more general mechanism for handling this.

I would propose the addition of something like keys `configFiles` and `secretFiles` added to `values.yaml` and those should be keys for objects with their own keyed attributes. The values of these can be file content. i.e. Something like this:

```
configFiles:
  someArbitraryKey: |
    This is the content of some file.
secretFiles:
  someOtherArbitraryKey: |
    This is the content of some other file.
```

You can then add one new `ConfigMap` and one new `Secret` resource to the templates and those should iterate over `configFiles` and `secretFiles`, respectively to incorporate their k=v pairs.

Both of the above can be unconditionally mounted to someplace in each pod. For instance, `/var/configs` and `/var/secrets` (maybe there's a better place; I'm just using my imagination right now).

For your specific use case, this _generic_ mechanism would allow you to specify the content of the service account file and when configuring the DNS challenge to use Google Cloud as the provider, you would use the `GCE_SERVICE_ACCOUNT_FILE` as normal to reference the location of the file, which would be `/var/secrets/<key>`.

Again-- I'd favor this more generic approach because I think it would get us out of more jams in the future.

jacobsvante commented:
@krancour changed as per your suggestion.

Maybe I should edit title and description to reflect the new generic functionality?

krancour commented on stable/traefik/templates/config-files.yaml:
```diff
@@ -0,0 +1,15 @@
+    heritage: "{{ .Release.Service }}"
+data:
+{{- range $filename, $fileContents := .Values.configFiles }}
+  {{ $filename }}: {{ $fileContents | b64enc | quote }}
```
Willing to be corrected if I am wrong, but I am pretty sure that values in a config map aren't supposed to be base64 encoded.

jacobsvante replied:
Thanks. Didn't test that one

krancour commented on stable/traefik/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: traefik
-version: 1.31.1
+version: 1.33.0
```
If this were merged next, version should be 1.32.0.

krancour commented:
Nice improvements in that last commit @jmagnusson. Thanks. See comment about base64 in config map.

jacobsvante commented:
@krancour Removed the `base64` of configmaps. Lgtm?

jameswinegar commented:
Just for prosperity. We just tested with for one of our projects. Works well.

jacobsvante committed:
Ability to add arbitrary Secrets and ConfigMaps

Google Cloud DNS is configured differently from all other ACME DNS providers in that it expects a file to be available inside the container (because of how the underlying lego library is set up). This didn’t map well to how helm packages are expected to be provided with configuration/secrets.

To support this the ability to specify any number of Secrets and ConfigMaps from the values.yaml file has been added.

Signed-off-by: Jacob Magnusson <m@jacobian.se>
